File: frontend/src/api/orgs.ts
import { apiFetch } from './client';

export interface OrgResponse {
  id: string;
  name: string;
  plan: string;
  created_at: string;
}

export interface MemberResponse {
  user_id: string;
  email: string;
  role: string;
  joined_at: string;
}

export interface JoinRequestResponse {
  id: string;
  user_id: string;
  user_email: string;
  status: string;
  created_at: string;
}

export function createOrg(name: string) {
  return apiFetch<OrgResponse>('/orgs', {
    method: 'POST',
    body: JSON.stringify({ name }),
  });
}

export function searchOrgs(query: string) {
  return apiFetch<{ orgs: OrgResponse[] }>(`/orgs/search?q=${encodeURIComponent(query)}`);
}

export function getCurrentOrg() {
  return apiFetch<OrgResponse>('/orgs/current');
}

export function refreshToken() {
  return apiFetch<{ token: string }>('/orgs/current/refresh-token', { method: 'POST' });
}

export function getMembers() {
  return apiFetch<{ members: MemberResponse[] }>('/orgs/current/members');
}

export function updateMemberRole(userId: string, role: 'owner' | 'member') {
  return apiFetch<MemberResponse>(`/orgs/current/members/${userId}`, {
    method: 'PATCH',
    body: JSON.stringify({ role }),
  });
}

export function removeMember(userId: string) {
  return apiFetch<void>(`/orgs/current/members/${userId}`, { method: 'DELETE' });
}

export function requestToJoin(orgId: string) {
  return apiFetch<JoinRequestResponse>(`/orgs/${orgId}/join`, { method: 'POST' });
}

export function getJoinRequests() {
  return apiFetch<{ requests: JoinRequestResponse[] }>('/orgs/current/join-requests');
}

export function resolveJoinRequest(requestId: string, action: 'accept' | 'decline') {
  return apiFetch<JoinRequestResponse>(`/orgs/current/join-requests/${requestId}`, {
    method: 'POST',
    body: JSON.stringify({ action }),
  });
}

File: api/src/api/schemas/orgs.py
"""Request/response schemas for organization management."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel


class OrgCreateRequest(BaseModel):
    """Request body for creating an organization."""

    name: str


class OrgResponse(BaseModel):
    """Organization info."""

    id: str
    name: str
    plan: str
    created_at: datetime


class OrgSearchResponse(BaseModel):
    """List of organizations from search."""

    orgs: list[OrgResponse]


class MemberResponse(BaseModel):
    """Organization member info."""

    user_id: str
    email: str
    role: str
    joined_at: datetime


class MembersListResponse(BaseModel):
    """List of organization members."""

    members: list[MemberResponse]


class JoinRequestResponse(BaseModel):
    """Join request info."""

    id: str
    user_id: str
    user_email: str
    status: str
    created_at: datetime


class JoinRequestsListResponse(BaseModel):
    """List of join requests."""

    requests: list[JoinRequestResponse]


class JoinRequestAction(BaseModel):
    """Action to take on a join request."""

    action: Literal["accept", "decline"]


class RoleUpdateRequest(BaseModel):
    """Request body for updating a member's role."""

    role: Literal["owner", "member"]


class TokenResponse(BaseModel):
    """Response containing a JWT token."""

    token: str
